File: upload_test.go
package main

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"testing"
	"time"
)

func TestExists(t *testing.T) {
	var gotSha string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotSha = r.URL.Query().Get("sha256")
		json.NewEncoder(w).Encode(map[string]bool{"exists": true})
	}))
	defer srv.Close()

	exists, err := NewClient(srv.URL, "test").Exists("abc123")
	if err != nil || !exists {
		t.Fatalf("exists=%v err=%v", exists, err)
	}
	if gotSha != "abc123" {
		t.Errorf("sha query param = %q", gotSha)
	}
}

func TestUploadOutcomes(t *testing.T) {
	cases := []struct {
		status int
		body   string
		want   OutcomeKind
	}{
		{200, `{"ok":true,"message":"Replay accepted — profiles are live now."}`, Accepted},
		{409, `{"message":"This exact replay file is already ingested."}`, Duplicate},
		{429, `{"message":"Too many uploads — try again later."}`, RateLimited},
		{400, `{"message":"Not a readable StarCraft II replay."}`, Rejected},
		{413, `{"message":"Replay too large."}`, Rejected},
		{500, `boom`, Transient},
	}
	for _, c := range cases {
		srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if err := r.ParseMultipartForm(32 << 20); err != nil {
				t.Errorf("not a multipart form: %v", err)
			}
			f, _, err := r.FormFile("replay")
			if err != nil {
				t.Errorf(`no "replay" form field: %v`, err)
			} else {
				raw, _ := io.ReadAll(f)
				if string(raw) != "replaydata" {
					t.Errorf("field bytes = %q", raw)
				}
			}
			w.WriteHeader(c.status)
			w.Write([]byte(c.body))
		}))
		out := NewClient(srv.URL, "test").Upload("x.SC2Replay", []byte("replaydata"))
		if out.Kind != c.want {
			t.Errorf("status %d: outcome %v, want %v (msg %q)", c.status, out.Kind, c.want, out.Message)
		}
		if c.status == 400 && out.Message != "Not a readable StarCraft II replay." {
			t.Errorf("message extraction failed: %q", out.Message)
		}
		srv.Close()
	}
}

func TestReadyCount(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`{"me":false,"until":null,"players":[{"battletag":"Foo#123"},{"battletag":"Bar#456"}]}`))
	}))
	defer srv.Close()
	n, names, err := NewClient(srv.URL, "test").ReadyCount()
	if err != nil || n != 2 || names[0] != "Foo#123" {
		t.Fatalf("n=%d names=%v err=%v", n, names, err)
	}
}

// TestWatcherOnce runs the full pipeline against a fake server: settle
// detection, MPQ sniff, sha pre-check, upload, state record.
func TestWatcherOnce(t *testing.T) {
	fixture, err := os.ReadFile(filepath.Join("testdata", "20260723-1808.SC2Replay"))
	if err != nil {
		t.Fatal(err)
	}
	dir := t.TempDir()
	replay := filepath.Join(dir, "Undead Assault reborn.SC2Replay")
	junk := filepath.Join(dir, "Other Map.SC2Replay")
	os.WriteFile(replay, fixture, 0o644)
	os.WriteFile(junk, []byte("not a real replay"), 0o644)
	old := time.Now().Add(-time.Minute)
	os.Chtimes(replay, old, old)
	os.Chtimes(junk, old, old)

	var posts int
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method == http.MethodGet {
			json.NewEncoder(w).Encode(map[string]bool{"exists": false})
			return
		}
		posts++
		w.Write([]byte(`{"ok":true,"message":"accepted"}`))
	}))
	defer srv.Close()

	state := LoadState(filepath.Join(dir, "state.json"))
	cfg := Config{Server: srv.URL, Dirs: []string{dir}, Once: true, PostSpacing: time.Millisecond}
	w := NewWatcher(cfg, NewClient(srv.URL, "test"), state, &consoleUI{})
	done := make(chan struct{})
	go func() {
		w.Run(context.Background())
		close(done)
	}()
	select {
	case <-done:
	case <-time.After(30 * time.Second):
		t.Fatal("watcher -once did not finish")
	}

	if posts != 1 {
		t.Errorf("expected exactly 1 upload POST, got %d", posts)
	}
	if rec := state.Files[replay]; rec.Status != "done" || rec.Reason != "uploaded" {
		t.Errorf("replay record = %+v", rec)
	}
	if rec := state.Files[junk]; rec.Status != "skip" {
		t.Errorf("junk record = %+v", rec)
	}

	// second run: nothing new, no further posts
	w2 := NewWatcher(cfg, NewClient(srv.URL, "test"), LoadState(filepath.Join(dir, "state.json")), &consoleUI{})
	w2.Run(context.Background())
	if posts != 1 {
		t.Errorf("re-run must not re-upload; posts=%d", posts)
	}
}

File: tests/upload.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { createServer, type Server } from 'node:http';
import { Client } from '../src/core/upload.ts';

function serve(
	handler: (req: import('node:http').IncomingMessage, body: Buffer) => { status: number; body: string }
): Promise<{ server: Server; url: string }> {
	return new Promise((resolve) => {
		const server = createServer((req, res) => {
			const chunks: Buffer[] = [];
			req.on('data', (c) => chunks.push(c));
			req.on('end', () => {
				const out = handler(req, Buffer.concat(chunks));
				res.statusCode = out.status;
				res.setHeader('content-type', 'application/json');
				res.end(out.body);
			});
		});
		server.listen(0, '127.0.0.1', () => {
			const addr = server.address() as { port: number };
			resolve({ server, url: `http://127.0.0.1:${addr.port}` });
		});
	});
}

test('exists passes the sha and reads the answer', async () => {
	let gotUrl = '';
	const { server, url } = await serve((req) => {
		gotUrl = req.url ?? '';
		return { status: 200, body: '{"exists":true}' };
	});
	try {
		assert.equal(await new Client(url, 'test').exists('abc123'), true);
		assert.equal(gotUrl, '/api/replays?sha256=abc123');
	} finally {
		server.close();
	}
});

test('upload maps statuses to outcomes', async () => {
	const cases: [number, string, string][] = [
		[200, '{"ok":true,"message":"Replay accepted — profiles are live now."}', 'accepted'],
		[409, '{"message":"This exact replay file is already ingested."}', 'duplicate'],
		[429, '{"message":"Too many uploads — try again later."}', 'ratelimited'],
		[400, '{"message":"Not a readable StarCraft II replay."}', 'rejected'],
		[413, '{"message":"Replay too large."}', 'rejected'],
		[500, 'boom', 'transient']
	];
	for (const [status, body, want] of cases) {
		let gotBody: Buffer = Buffer.alloc(0);
		let contentType = '';
		const { server, url } = await serve((req, reqBody) => {
			gotBody = reqBody;
			contentType = String(req.headers['content-type']);
			return { status, body };
		});
		try {
			const out = await new Client(url, 'test').upload('x.SC2Replay', Buffer.from('replaydata'));
			assert.equal(out.kind, want, `status ${status}`);
			assert.match(contentType, /^multipart\/form-data/);
			assert.ok(gotBody.includes(Buffer.from('name="replay"')), 'multipart field name');
			assert.ok(gotBody.includes(Buffer.from('replaydata')), 'file bytes present');
			if (status === 400) assert.equal(out.message, 'Not a readable StarCraft II replay.');
		} finally {
			server.close();
		}
	}
});

test('ready returns count, battletags and expiries', async () => {
	const { server, url } = await serve(() => ({
		status: 200,
		body: '{"me":false,"until":null,"players":[{"battletag":"Foo#123","until":"2026-01-01T01:00:00Z"},{"battletag":"Bar#456","until":"2026-01-01T02:00:00Z"}]}'
	}));
	try {
		const r = await new Client(url, 'test').ready();
		assert.deepEqual(r, {
			count: 2,
			names: ['Foo#123', 'Bar#456'],
			players: [
				{ battletag: 'Foo#123', until: '2026-01-01T01:00:00Z' },
				{ battletag: 'Bar#456', until: '2026-01-01T02:00:00Z' }
			]
		});
	} finally {
		server.close();
	}
});
